lib/commands/web.js

```diff
@@ -22,58 +28,91 @@ const getSafariIsIphone = _.memoize(async function getSafariIsIphone (sessionId,
   // sessionId parameter is for memoizing per session
   let isIphone = true;
   try {
-    const useragent = await driver.execute('return navigator.userAgent');
-    isIphone = useragent.toLowerCase().includes('iphone');
+    const userAgent = await driver.execute('return navigator.userAgent');
+    isIphone = userAgent.toLowerCase().includes('iphone');
   } catch (err) {
     log.warn(`Unable to find device type from useragent. Assuming iPhone`);
     log.debug(`Error: ${err.message}`);
   }
   return isIphone;
 });
 
-extensions.getElementHeightMemoized = _.memoize(async function (key, el) {
+const getSafariIsIphoneX = _.memoize(async function getSafariIsIphone (sessionId, driver) {
+  // sessionId parameter is for memoizing per session
+  // because of the way lodash memoization works, this function cannot call
+  // `getSafariIsIphone` and short-circuit when not on an iPhone
+  let isIphoneX = false;
+  try {
+    const height = await driver.execute('return window.screen.availHeight');
+    const width = await driver.execute('return window.screen.availWidth');
+    // check for the correct height and width
+    isIphoneX = (height === IPHONE_X_HEIGHT && width === IPHONE_X_WIDTH) ||
+                (height === IPHONE_X_WIDTH && width === IPHONE_X_HEIGHT);
+  } catch (err) {
+    log.warn(`Unable to find device type from useragent. Assuming not iPhone X`);
+    log.debug(`Error: ${err.message}`);
+    isIphoneX = false;
+  }
+  return isIphoneX;
+});
+
+const getElementHeightMemoized = _.memoize(async function (key, driver, el) {
   el = util.unwrapElement(el);
-  return (await this.getNativeRect(el)).height;
+  return (await driver.getNativeRect(el)).height;
 });
 
-extensions.getExtraTranslateWebCoordsOffset = async function () {
+extensions.getExtraTranslateWebCoordsOffset = async function (coords, webviewRect) {
   let offset = 0;
 
   // keep track of implicit wait, and set locally to 0
   const implicitWaitMs = this.implicitWaitMs;
 
-  // try to see if there has been scrolling
   try {
     this.setImplicitWait(0);
 
+    // check if the full url bar is up
     await this.findNativeElementOrElements('accessibility id', 'ReloadButton', false);
     // reload button found, which means scrolling has not happened
-    offset += await getSafariIsIphone(this.opts.sessionId, this) ?
-      IPHONE_EXTRA_WEB_COORD_NON_SCROLL_OFFSET:
-      IPAD_EXTRA_WEB_COORD_NON_SCROLL_OFFSET;
+    if (await getSafariIsIphoneX(this.opts.sessionId, this)) {
+      offset += IPHONE_X_EXTRA_WEB_COORD_NON_SCROLL_OFFSET;
+    } else if (await getSafariIsIphone(this.opts.sessionId, this)) {
```

Comment:
can we optimise this? this block of code will make at least 6 calls to the remote end if the device is ipad

Reply:
yeah, maybe there's a way to make one `driver.execute` call which returns an object full of all the info we might need